### PR TITLE
rfc: Introduce the FmtPrometheus and FmtLabels traits

### PR DESCRIPTION
**Pull request by olix0r**
There are many different types of things that implement `fmt::Display`
in the metrics library: lists of key-value labels, individial label keys
or values, groups of metrics output, etc. While modifying metrics code,
it can be very easy to use something that's not a label in the context
of a label.

### CODE REVIEW

**Comment by olix0r** [comment]
https://github.com/linkerd/linkerd2-proxy/pull/68 shows why this should be useful -- we only want/need to store a `ctx::Proxy` type in our data structures. The label-formatting details should be decoupled from the structural representation of the metrics store.

File: src/telemetry/metrics/labels.rs [hunk 1]
```diff
@@ -147,21 +156,25 @@ impl ResponseLabels {
...
-        )?;
-
-        if let Some(ref status) = self.grpc_status_code {
-            // leading comma added between the status code label and the
-            // gRPC status code labels, if there is a gRPC status code.
-            write!(f, ",grpc_status_code=\"{}\"", status)?;
-        }
+impl FmtLabels for ResponseLabels {
+    fn fmt_labels(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let l = (((&self.request_labels,
```
Comment (hawkw) [inline src/telemetry/metrics/labels.rs]: Personally, I find this expression rather hard to parse...

File: src/telemetry/metrics/prom.rs [hunk 2]
```diff
@@ -0,0 +1,126 @@
+use std::fmt;
+use std::marker::PhantomData;
+
+/// Writes a block of metrics in prometheus-formatted output.
+pub trait FmtPrometheus {
```
Comment (hawkw) [inline src/telemetry/metrics/prom.rs]: I feel like the `FmtPrometheus` and `FmtMetric` trait names aren't super clear --- the comments help a bit, but just seeing the trait names, it's not immediately clear which does what. It might be good if that could be made clearer...

File: src/telemetry/metrics/serve.rs [hunk 3]
```diff
@@ -180,3 +182,9 @@ impl Error for ServeError {
         }
     }
 }
+
+impl<F: FmtPrometheus> fmt::Display for PromWriter<F> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
```
Comment (hawkw) [inline src/telemetry/metrics/serve.rs]: This layer of indirection seemed unnecessary to me until I realized you couldn't `fmt::Display::fmt(metrics, &mut writer)` because `writer` is `io::Write` but not a `Formatter`. Maybe there ought to be a comment here somewhere?

File: src/telemetry/metrics/prom.rs [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+// ===== impl FmtLabels =====
+
+impl<'a, A: FmtLabels + 'a> FmtLabels for &'a A {
+    fn fmt_labels(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        (*self).fmt_labels(f)
+    }
+}
+
+impl<A: FmtLabels, B: FmtLabels> FmtLabels for (A, B) {
```
Comment (hawkw) [inline src/telemetry/metrics/prom.rs]: The helper methods on tuples are cute. I wonder if it would be simpler to just define one on `IntoIterator<Item=&'a FmtLabels>`, though? Then, the case of `Option`al labels could be handled with `filter_map`.

Comment (olix0r) [reply src/telemetry/metrics/prom.rs]: we need to know that there is at least one valid label, otherwise all of the comma insertion breaks.  This is why we dont provide impls for `Option<A>` or `(Option<A>, Option<B>)`...

are you suggesting that we support `(A, IntoIterator<Item=FmtLabels>)` instead of the Option? would that work?

Comment (hawkw) [reply src/telemetry/metrics/prom.rs]: > we need to know that there is at least one valid label, otherwise all of the comma insertion breaks. 

I think we can do something like this:
```rust 
let mut iter = i.into_iter() // iterator of Option<FmtLabels>
    .filter_map(|i| i); // filter out any nones.

if let Some(label) = iter.next() {
    write!(f, "{}", label)?;

    for label in iter {
        write!(f, ",{}", label)?;
    }
}
```

regardless, this seems fine as-is!

Comment (olix0r) [reply src/telemetry/metrics/prom.rs]: I think this has two problems:

1. `FmtLabels` must not represent an empty set of labels. Either that, or the `FmtLabels` type would need to expose a `is_empty()`. This is important for the purposes of composition.  If I have an iterator of `FmtLabels` that i want to combine, how do i know if any one of those items is itself empty?
2. AIUI, we would almost always need to use `Box<FmtLabels>`, because we are trying to handle diverse types.
